### ISSUE TITLE
sim pauses when large numbers of bunnies are born

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/natural-selection#60]
Related to https://github.com/phetsims/tandem/issues/173 (PhetioGroup is not scalable).

As the number of bunnies increases, we start to experience problems with responsiveness and performance.  This is even before starting to collect the large amounts of data mentioned in #46. 

When large numbers of bunnies mate at 12:00 on the generation clock, that results in many calls to `PhetioGroup.createNextElement`.  For example, if I have 800 bunnies, that's 400 pairs x 4 per litter = 1600 bunnies that will be created, and 1600 calls to `PhetioGroup.createNextElement`. This results in a very noticeable pause in animation - seconds in some cases.

Is this pause in animation acceptable?

**Comment by pixelzoom** [comment]
This problem can be demonstrated with https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.31/phet/natural-selection_en_phet.html?secondsPerGeneration=1

1. Go to the Lab screen
2. Press the "Add a Mate" button
3. Observe how the pause in animation increases with each successive generation.

**Comment by pixelzoom** [comment]
@samreid determined that 58% of time is in allocating BunnyNode, 31% is in allocating the PressListener for each BunnyNode (used to select bunnies).  See profiler screenshot below.   He suggested pre-allocating 3000 BunnyNode instances.  Since we don't know what a BunnyNode will look like until a Bunny is born, BunnyNode would need to be fully mutable. 

![image](https://user-images.githubusercontent.com/3046552/80150368-80651f00-8575-11ea-8bfb-a5b4da0eea0f.png)

**Comment by samreid** [comment]
I also mentioned on slack:

> If you preallocate all Bunny and BunnyNode now before studio is optimized, it would probably make it unusable.???

**Comment by pixelzoom** [comment]
Notes on computations of how many Bunny instances are needed.

**Live bunnies**: Bunnies take over the world when the population is >= 1000.  So the maximum number of live bunnies occurs when 998 bunnies mate.  998 live bunnies + 998/2 pairs * 4 bunnies per litter = 2994 live bunnies.

**Dead bunnies**: For the selected bunny, the Pedigree graph will display 4 generations of ancestors, which is 30 bunnies, possibly all dead.   Bunnies share ancestors, so we don't need 2994 * 30 dead bunnies.  Bunnies live for 2 generations, then die of old age.  So disposing of all bunnies who were born >= 6 generations ago might be a reasonable strategy.  In any case, I'm not sure how to accurately compute the maximum number of dead bunnies.

**Comment by pixelzoom** [comment]
More discussion on total number of bunnies with @samreid on Slack.  We think the total number of Bunny instances maybe be ~4000 per screen. 

Sam Reid  4:06 PM
So a loose upper bound for number would be 4000, since you need to keep dead bunnies <=4 generations away?

Chris Malley  4:07 PM
I don't follow. 4 generations is 30 ancestors per bunny.

Sam Reid  4:28 PM
Right but generation N-1 has a max of 1000 bunnies, N-2 has max 1000 bunnies, N-3 has max 1000 bunnies, N-4 has max 1000 bunnies, and you said “I need to keep all dead bunnies who are <= 4 generations away from a living bunny.”  So the max amount of dead bunnies you need to keep is 4000, right?

Chris Malley  4:28 PM
I see. Yes, that would probably work.

**Comment by pixelzoom** [comment]
More thoughts on pre-allocation, after discussing with @samreid via Zoom ...

Pre-allocating would mean that PhetioGroup is unnecessary.  That represents a lot of work here, but at least it has informed PhetioGroup evolution.

Pre-allocating BunnyNode instances seems doable.  It's currently a matter of associating a new Bunny instance with the BunyNode, and changing the BunnyNode's image based on the phenotype of the Bunny instance.  It will become a bit more complicated for the BunnyNode instances shown in the Pedigree graph, since they show more info, but maybe they should be a subclass of BunnyNode (with no PressListener) and created on demand. 

Pre-allocating Bunny instances seems potentially problematic, because:  

* Studio (and the sim) would need to be able to start up with 4000 Bunny instances per screen allocated.
* Bunny is a complicated data structure (see outline below), so mutation would not be trivial.
* Bunny mutation will modify instrumented Properties, and that will result in data stream messages. Is that OK? Or do we need a way to turn off data stream while an element from a pool is being mutated?
* Bunny mutation would involve the need to reinstantiate or mutate instances of other classes, specifically Genotype, GenePair, and Phenotype.

```
Bunny

  number generation
  Bunny father
  Bunny mother

  Property.<Vector3> positionProperty
  EnumerationProperty.<SpriteDirection> directionProperty
  NumberProperty ageProperty
  BooleanProperty isAliveProperty

  // genetic blueprint, inherited from parents
  Genotype genotype 
    GenePair furPair
      Allele fatherAllele
      Allele motherAllele
    GenePair earsPair
      Allele fatherAllele
      Allele motherAllele
    GenePair teethPair
      Allele fatherAllele
      Allele motherAllele

  // appearance, based on genotype and allele dominance
  Phenotype phenotype 
    Allele furAllele
    Allele earsAllele
    Allele teethAllele
  ```

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/natural-selection/issues/60#issuecomment-618677008, @samreid noted that PressListener allocation for each BunnyNode was a significant cost.  In the above commit, I'm now using a single PressListener for the entire collection of BunnyNodes, rather than one per BunnyNode.

**Comment by pixelzoom** [comment]
The build version is now feeling pretty responsive.

**Comment by pixelzoom** [comment]
@samreid and I looked at this briefly today via Zoom.  ~10% is spent creating the selection Rectangle in BunnyNode.  It's currently part of BunnyNode because the selection Rectangle needs to move and scale with BunnyNode.  But perhaps I can find a way to use a single selection Rectangle for the entire BunnyNode collection.

**Comment by pixelzoom** [comment]
I tried creating the selection Rectangle on demand, here's the patch:

<details>
<summary>Patch</summary>

```diff
Index: js/common/model/BunnyCollection.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/model/BunnyCollection.js	(revision 507a91db0f1d6261bac156e8a712ecf9a2a80d01)
+++ js/common/model/BunnyCollection.js	(date 1588718689922)
@@ -268,27 +268,29 @@
     }
 
     // Mate adjacent pairs from the collection, applying mutations where appropriate.
-    for ( let i = 1; i < bunnies.length; i = i + 2 ) {
+    NaturalSelectionUtils.logTime( 'mate', () => {
+      for ( let i = 1; i < bunnies.length; i = i + 2 ) {
 
-      const father = bunnies[ i ];
-      const mother = bunnies[ i - 1 ];
+        const father = bunnies[ i ];
+        const mother = bunnies[ i - 1 ];
 
-      for ( let j = 0; j < NaturalSelectionConstants.LITTER_SIZE; j++ ) {
+        for ( let j = 0; j < NaturalSelectionConstants.LITTER_SIZE; j++ ) {
 
-        this.createBunny( {
-          father: father,
-          mother: mother,
-          generation: generation,
-          genotypeOptions: {
-            furMutation: ( furIndices && furIndices.indexOf( bornCount ) >= 0 ) ? furMutation : null,
-            earsMutation: ( earsIndices && earsIndices.indexOf( bornCount ) >= 0 ) ? earsMutation : null,
-            teethMutation: ( teethIndices && teethIndices.indexOf( bornCount ) >= 0 ) ? teethMutation : null
-          }
-        } );
+          this.createBunny( {
+            father: father,
+            mother: mother,
+            generation: generation,
+            genotypeOptions: {
+              furMutation: ( furIndices && furIndices.indexOf( bornCount ) >= 0 ) ? furMutation : null,
+              earsMutation: ( earsIndices && earsIndices.indexOf( bornCount ) >= 0 ) ? earsMutation : null,
+              teethMutation: ( teethIndices && teethIndices.indexOf( bornCount ) >= 0 ) ? teethMutation : null
+            }
+          } );
 
-        bornCount++;
-      }
-    }
+          bornCount++;
+        }
+      }
+    } );
 
     assert && assert( bornCount === numberToBeBorn, 'unexpected number of bunnies were born' );
     phet.log && phet.log( `${bornCount} bunnies born` );
Index: js/common/view/BunnyNode.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/view/BunnyNode.js	(revision 507a91db0f1d6261bac156e8a712ecf9a2a80d01)
+++ js/common/view/BunnyNode.js	(date 1588719237750)
@@ -50,18 +50,8 @@
       bottom: 0
     } );
 
-    // Rectangle that appears around this Node when bunny is selected
-    const selectionRectangle = new Rectangle( wrappedImage.bounds.dilated( 5 ), {
-      fill: 'rgba( 0, 0, 0, 0.25 )',
-      stroke: 'blue',
-      lineWidth: 2.5,
-      cornerRadius: NaturalSelectionConstants.CORNER_RADIUS,
-      center: wrappedImage.center,
-      pickable: false
-    } );
-
     assert && assert( !options.children, 'BunnyNode sets children' );
-    options.children = [ selectionRectangle, wrappedImage ];
+    options.children = [ wrappedImage ];
 
     // Red dot at the origin
     if ( NaturalSelectionConstants.SHOW_ORIGIN ) {
@@ -70,11 +60,25 @@
 
     super( bunny, options );
 
+    // Rectangle that appears around this Node when bunny is selected. Created on demand.
+    let selectionRectangle = null;
+
     // Indicate that this bunny is selected
     const selectedBunnyListener = someBunny => {
+      if ( !selectionRectangle ) {
+        selectionRectangle = new Rectangle( wrappedImage.bounds.dilated( 5 ), {
+          fill: 'rgba( 0, 0, 0, 0.25 )',
+          stroke: 'blue',
+          lineWidth: 2.5,
+          cornerRadius: NaturalSelectionConstants.CORNER_RADIUS,
+          center: wrappedImage.center,
+          pickable: false
+        } );
+        this.insertChild( 0, selectionRectangle );
+      }
       selectionRectangle.visible = ( someBunny === bunny );
     };
-    selectedBunnyProperty.link( selectedBunnyListener );
+    selectedBunnyProperty.lazyLink( selectedBunnyListener );
 
     // @private
     this.disposeBunnyNode = () => {
```

</details> 

Here's the performance comparison before and after, measuring the `for` loop in BunnyNode.mateAllBunnies using `NaturalSelectionUtils.logTime`:

| generation | bunnies born | before (ms) | after (ms) |
| :--: | :--: | :--: | :-: |
| 1 | 4 | 8 | 7 |
| 2 | 12 | 21 | 17 |
| 3 | 36 | 60 | 51 |
| 4 | 108 | 144  | 119 |
| 5 | 320 | 360 | 402 |
| 6 | 952 | 924 | 967 |

Note that the patch actually take _more_ time for larger numbers of bunnies. My guess is that creating and lazyLink'ing the input listener is taking more time than is saved by not creating the selection Rectangle.

**Comment by pixelzoom** [comment]
Labeling for design meeting, to get input from team on what acceptable responsiveness is.

**Comment by pixelzoom** [comment]
Current pause would be acceptable to publish, but still worth investigating if there are opportunities to make it more efficient. 

@ariel-phet suggested having a "standard minimum pause" when bunnies are born, but consensus was that we don't need that.

**Comment by pixelzoom** [comment]
Most of the sim is now fleshed out, so this can be revisited.

For this scenario:

1. start sim, go to Lab screen
2. press "Add a Mate" button
3. wait for bunnies to take over the world

... here's the performance of the unbuilt version (with assertions enabled) vs the built version:

| generation | bunnies born | unbuilt (ms) | built (ms) |
| :--: | :--: | :--: | :-: |
| 1 | 4 | 9.6 | 4.7 |
| 2 | 12 | 20.9 | 11.4 |
| 3 | 36 | 58.3 | 30.4 |
| 4 | 108 | 134.8  | 68.6 |
| 5 | 320 | 430.9 | 163.4 |
| 6 | 952 | 892.4 | 410.5 |

The unbuilt version with assertions enabled is about 2x slower, because of the time required to check assertion expressions. The unbuilt version with assertions _disabled_ is about the same as the built version, just a tad slower.

Less than 1/2 second seems acceptable to me, and https://github.com/phetsims/natural-selection/issues/60#issuecomment-625458324 indicated that we were willing to publish with as much as 1 second paused.   So I'm inclined to close this issue, but I'll leave that up to @amanda-phet.  If the current performance is OK, feel free to close this issue.  If you'd like me to try speeding things up more, assign back to me, let me know what an acceptable delay is, and I'll be happy to investigate.

**Comment by pixelzoom** [comment]
Labeled for design meeting to get feedback on what an acceptable delay is.

**Comment by pixelzoom** [comment]
7/2/2020 design meeting:

Under 500 ms is acceptable.

**Comment by pixelzoom** [comment]
> Check on an older iPad or older MacBook Pro.

In [1.0.0-dev.61](https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.61/phet/natural-selection_en_phet.html), I added a temporary display that shows the time for each mating cycle.  For example:

<img width="503" alt="screenshot_419" src="https://user-images.githubusercontent.com/3046552/86496040-d35b0f80-bd38-11ea-8b9e-151f9db18d1d.png">

To test, I started the sim, went to the Lab screen, pressed the "Add a Mate" button, and waited for bunnies to take over the world.  I did 3 runs on each platforms.  Here are the results for the 6th generation, where 952 bunnies are born:

| platform | 3 tests (ms) | average (ms) |
| :--- | ---: | ---: |
| iPad6 + iOS 13.5.1 + Safari  | 757, 1040, 1014 | 937 |
| iPad2 + iOS 9.3.5 + Safari | 4105, 3496, 3263 | 3621 |
| 2009 MacBook Pro + macOS 10.11.6 + Chrome 83 | 1455, 1116, 1158 | 1243 |

As you can see, the newer iPad6 with latest iOS was borderline acceptable -- it averaged under 1 second, but was sometime over 1 second.

The older platforms were well over 1 second.  These are admittedly very old platforms, with the iPad2 not even supported by PhET. But they are what I had available, and are probably typical of other older (supported) platforms.

There are no clear/straightforward ways to speed this sim up.  So before I dive in, it might be good some data for PhET test devices.  @ariel-phet is QA available to do this?

**Comment by pixelzoom** [comment]
In the above commits, I added a `showTimeToMate` query parameter.   So this debug feature will be available (but hidden) in future versions. 

I also created the bunny selection rectangle (in BunnyNode.js) on demand, which reduced the delay significantly as follows (values are averages over 3 runs):

| platform | before (ms) | after (ms) |
| :--- | ---: | ---: |
2019 MacBook Pro | 410 | 326 |
2009 MacBook Pro 965 | 1243 | 965 |
iPad6 | 937 | 522 |

I didn't test this on iPad2.

**Comment by pixelzoom** [comment]
Just for kicks, I tried eliminating the PhetioGroup that created BunnyNode instances, since it's unclear whether BunnyNode really needs to be instrumented.   On my 2019 MacBook Pro, the delay reduction was negligible - from 326 ms to 317 ms.

**Comment by ariel-phet** [comment]
@pixelzoom it seems like what you are asking is a very quick test. Please contact @KatieWoe directly on slack and see if she can squeeze this in with some other testing (I imagine a QA person working on another sim on a specific platform could just do this test quickly)

**Comment by pixelzoom** [comment]
I messaged @KatieWoe on Slack, and she can fit this in sometime this week.   Here are the instructions:

Please test with [1.0.0-dev.62](https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.62/phet/natural-selection_en_phet.html) with a range of PhET test devices, including performance-challenged devices.  Use the steps and table below, which will record 3 times and an average. 

Steps:
1. Record the test device name and browser+version in the table below.
1. Run the sim with `?showTimeToMate&secondsPerGeneration=1`
2. Go to the _Lab_ screen.
3. Press the "Add a Mate" button.
4. Wait for the "Bunnies have taken over the world" dialog to appear, then close it.
5. In the table, record the "time to mate" value that appears in the upper-left corner of the screen.
6. Press the "Start Over" button, press the "Add a Mate" button.
7. Wait for the "Bunnies have taken over the world" dialog to appear, then close it.
8. In the table, record the "time to mate" value that appears in the upper-left corner of the screen.
9. Press the "Start Over" button, press the "Add a Mate" button.
10. Wait for the "Bunnies have taken over the world" dialog to appear, then close it.
11. In the table, record the "time to mate" value that appears in the upper-left corner of the screen.
12. In the table, record the average of the 3 times.

| test device | browser + version | time1 (ms) | time2 (ms) | time3 (ms ) | average (ms ) |
|---|---| ---:| ---:|---:|---:|
| | | | | | |

**Comment by KatieWoe** [comment]
@pixelzoom the table is filled out with the platforms that felt most important. Let me know if you want any more specific platforms tested.

**Comment by pixelzoom** [comment]
Thanks @KatieWoe.

**Comment by pixelzoom** [comment]
@KatieWoe and I discussed the test results shown in https://github.com/phetsims/natural-selection/issues/60#issuecomment-654325501.  

* The "Dell Win 10" machine is her personal machine, and she said that Firefox is typically slower than Chrome and Edge. 

* "Lovelace - ChromeOS" is an older Chromebook, typically PhET's worst-performing test platform (and it didn't disappoint here :)  The sim felt generally usable, and the pause at generations 5 & 6 were a bit annoying, but not a huge deal. 

* The other platforms satisfied our desired < 1 sec pause.

So... I'm on the fence about what to do here.  Some options:

(1) Check performance with other Node `renderer` options.

(2) Preallocate BunnyNode instances, add them to scenegraph once, reuse them, make the ones that are in use visible. I'm not convinced that this will solve the performance problem, because BunnyNode image will need to be changed to match the phenotype.  This will also make for a very odd PhET-iO API, unless we uninstrument BunnyNode.

(3) Switch to Canvas for rendering the bunny images in the environment. Use 1 CanvasNode to draw all bunnies. This will make it impossible to PhET-iO-instrument individual bunnies in the view.  It will make selecting a bunny a lot more complicated.  And it means tossing the currently implementation of the bunny view and starting over.

**Comment by pixelzoom** [comment]
Running the PhET-iO standalone version results in an even larger delay.   On my 2019 MacBook Pro, the average of 3 runs (429, 465, 496 ms) is 463 ms of delay for generation 6.  Compared to the `brand=phet` version, where the delay was 326 ms, that a 42% increase in delay.  I'm not sure how that will translate to other platforms, but I'm guessing that it will at least that bad.

PhET-iO brand was run with 
`?brand=phet-io&phetioStandalone&showTimeToMate&secondsPerGeneration=1`

PhET brand was run with `?brand=phet&showTimeToMate&secondsPerGeneration=1`

**Comment by pixelzoom** [comment]
7/7/2020 I discussed performance with @jonathanolson.  We paired for ~1 hour, looking at performance in Chrome dev tools.

Summary:

* Most of the overhead is in Node.
* There is no single bottleneck to be relieved; it's cumulative overhead.
* Node overhead has increased due to PhET-iO (PhetioObject) and a11y (ParallelDOM).
* Performance of `brand=phet-io` is demonstrably worse (delay was ~5 seconds on @jonathanolson's MacBook!)
* Using `renderer='webgl'` resulted in a tiny improvement, and doesn't work with the bunny selection rectangle.
* Adding `preventFit: true` to the parent node for sprites (bunnies, wolves, shrubs) resulted in a tiny improvement.
* Rather than implement something sim-specific based on Canvas or WebGL, @jonathanolson recommends that we develop a something general for rendering high-performance sprites in scenery.

@jonathanolson please chime in if I missed anything.

Labeling for discussion at design meeting.

**Comment by kathy-phet** [comment]
Chris, How do I run this as PhET-iO? If I use the dev version with ?brand=phet-io&phetioStandalone&showTimeToMate&secondsPerGeneration=1 it only comes up with phet brand still. Do I have to run on phettest.

**Comment by pixelzoom** [comment]
@kathy-phet:

If you want to run a dev version...  I have been building all dev versions for brand=phet and brand=phet-io.  The phet-io version is in a hidden directory, password protected.  It's probably easiest is to navigate to the brand=phet version, then manually modify the URL in your browser. 

For example, the URL for latest dev version for brand=phet is:
https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.64/phet/natural-selection_en_phet.html

Change the URL to this, and you'll get the PhET-iO Index page:
https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.64/phet-io 

If you want to run master... Go to http://phettest.colorado.edu (VPN required). Click on the "natural-selection" link. Then manually change the query parameter to `?brand=phet-io&phetioStandalone` plus whatever sim-specific query parameters you want.

**Comment by pixelzoom** [comment]
7/8/2020 design meeting @amanda-phet @kathy-phet @ariel-phet 

We asked @jonathanolson about:
> Rather than implement something sim-specific based on Canvas or WebGL, @jonathanolson recommends investing in a general solution for rendering high-performance sprites in scenery.

@jonathanolson has some of this implemented already. @kathy-phet asked him to proceed, and track hours for Natural Selection.

I'm going to remove instrumentation of bunnies and wolves in the view (BunnyNodeCollection, WolfNodeCollection).

**Comment by pixelzoom** [comment]
Deferred for #128 (high-performance sprites).

**Comment by pixelzoom** [comment]
Conversion to high-performance sprites in #128 is completed.

I'll be doing more thorough performance tests. But with `?showTimeToMate` on my 2019 MacBook Pro + macOS 10.15.5 + Chrome 83, the time to create 952 bunnies at generation 6 has dropped from 326ms to 130ms

**Comment by pixelzoom** [comment]
@KatieWoe please re-run a few tests, following the Steps in https://github.com/phetsims/natural-selection/issues/60#issuecomment-654325501.

For brand=phet, run https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.72/phet/natural-selection_en_phet.html, and enter the browser versions and times (in ms) in this table:

| test device | browser + version | time<sub>1</sub> | time<sub>2</sub> | time<sub>3</sub> | average |
|---|---| ---:| ---:|---:|---:|
| Dell Win 10  (Katie's) | Chrome ??? |  | |  |  |
| Dell Win 10  (Katie's) | Firefox ??? | |  |  | |
| Lovelace - ChromeOS | ChromeOS ??? |  |  |  |  |
| Dirac - MacBook Pro macOS 10.12.6 | Safari 12.1.2 |  |  |  |  |

For brand=phet-io, run https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.72/phet-io/natural-selection_all_phet-io.html?postMessageOnError&phetioStandalone, and enter the browser versions and times (in ms) in this table:

| test device | browser + version | time<sub>1</sub> | time<sub>2</sub> | time<sub>3</sub> | average |
|---|---| ---:| ---:|---:|---:|
| Dell Win 10  (Katie's) | Chrome ??? |  | |  |  |
| Dell Win 10  (Katie's) | Firefox ??? | |  |  | |
| Lovelace - ChromeOS | ChromeOS ??? |  |  |  |  |
| Dirac - MacBook Pro macOS 10.12.6 | Safari 12.1.2 |  |  |  |  |

**Comment by KatieWoe** [comment]
Biggest lag seemed to be on last generation, more so than before. ChromeOS was inconsistent performance.
Larger problem was actually the Start Over button. It could take a few seconds on ChromeOS and was slower than expected on all platforms. I thought I had misclicked on ChromeOS and clicked again, and that seemed to register as Add a Mate, which seems problimatic.

**Comment by pixelzoom** [comment]
> Biggest lag seemed to be on last generation, more so than before.

@KatieWoe can you explain this comment? All of the recorded times are for the last generation. And they are all significantly lower than the times that you previously recorded in https://github.com/phetsims/natural-selection/issues/60#issuecomment-654325501.

> ChromeOS was inconsistent performance.

"Lovelace - ChromeOS" is not listed in the PhET Asset spreadsheet.  @KatieWoe reports via Slack that it's an Asus C300S.  I couldn't find specs on the C300S, but here are specs on the C300: https://www.asus.com/us/Laptops/ASUS_Chromebook_C300/specifications/

**Comment by KatieWoe** [comment]
I mean that the difference between the second to last generation and the last generation is more significant than it was in the last test.

**Comment by pixelzoom** [comment]
Via Zoom, @KatieWoe and I discussed her comments in https://github.com/phetsims/natural-selection/issues/60#issuecomment-662625708.  Summary:

* "Biggest lag seemed to be on last generation, more so than before":  This refers to the proportional increase been the 6th (last) and 5th generation. In the previous tests (1.0.0-dev.62), the time increased incrementally with each generation. In the latest test (1.0.0-dev.72), the first 5 generations all happened rather quickly, and there was a proportionally large increase at the 6th generation.  

* "Larger problem was actually the Start Over button. It could take a few seconds on ChromeOS and was slower than expected on all platforms."   This was a problem with brand=phet-io, the "Start Over" button did not feel responsive. For brand=phet, performance on all plaforms other than ChromeOS was acceptable, and didn't get worse from 1.0.0-dev.62 to 1.0.0-dev.72.

**Comment by pixelzoom** [comment]
I've created a new issue, #140, to investigate the performance of "Start Over".

For dev versions >= 1.0.0-dev.73, the query parameter is now `showTimes`.    It will show both the time to mate, and the time to perform 'Start Over', in the upper-left corner of the screen:

<img width="300" alt="screenshot_435" src="https://user-images.githubusercontent.com/3046552/88222719-99887500-cc23-11ea-9421-5f6108a12b48.png">

**Comment by pixelzoom** [comment]
Based on the table in https://github.com/phetsims/natural-selection/issues/60#issuecomment-662582884, there is a big performance hit across all platforms in `brand=phet-io`.  I'd like to work with someone on the PhET-iO team to identify bottlenecks that are specific to `brand=phet-io`.   And I'm wondering if bottlenecks that were identified for #140 (https://github.com/phetsims/axon/issues/316, https://github.com/phetsims/phet-io/issues/1689) are relevant here.

**Comment by samreid** [comment]
We pushed an improvement for https://github.com/phetsims/phet-io/issues/1689, does it speed up the "Start Over" time for this test?  Please reach out to me on slack to schedule a meetup time for further performance analysis.

**Comment by pixelzoom** [comment]
@samreid The performance improvement in phetsims/phet-io#1689 is on my radar, and I was hoping to test performance this past week.   But publishing a prototype has been given higher priority, and that prototype does not include PhET-iO.  So at this point I really don't know when I'll be getting to this.

**Comment by pixelzoom** [comment]
This issue is unrelated to the performance of "Start Over".  Moving the above question to https://github.com/phetsims/natural-selection/issues/140.

**Comment by samreid** [comment]
Unassigning until we schedule a collaboration time.

**Comment by pixelzoom** [comment]
This issue is a performance problem that occurs when large numbers of dynamic elements (bunnies) are created.  Which PhET-iO GitHub issues are relevant to this issue?  Or do we still need to investigate where the bottlenecks are?  Assigning to the iO team.

**Comment by samreid** [comment]
>  that were identified for #140 (phetsims/axon#316, phetsims/phet-io#1689) are relevant here.

Those are mostly for disposal.  I don't recall a recent performance investigation into creation of large numbers of bunnies.  Want to collaborate on that?

**Comment by pixelzoom** [comment]
Investigating in Chrome Dev Tools "Performance" tab...

1. Start the sim in Incognito with `?brand=phet-io&phetioStandalone` (no assertions)
2. Go to Lab screen
3. Press "Add a Mate" button
4. Just before the clock reaches Generation 6, start recording in the Performance tab.
5. Immediately after the "Bunnies have taken over..." dialog is displayed, stop recording in the Performance tab. 

Examining the Call Tree, there's a fork at `stepSimulation.execute.(anonymous)`:

<img width="1177" alt="screenshot_471" src="https://user-images.githubusercontent.com/3046552/89956988-9154a180-dbf3-11ea-8a65-1b6a5af445f9.png">

47.8% is spent in `stepSimulation.execute.(anonymous).step`. Drilling down, we find (as expected) that most of the time is spent in BunnyCollection.mateBunnies (see below).

<img width="1176" alt="screenshot_472" src="https://user-images.githubusercontent.com/3046552/89957165-f8725600-dbf3-11ea-8f79-396aa64600db.png">

Drilling down further, 28.5% is spent in BunnyGroup.createElement, see screenshot.  Drilling down further in here, I don't see any clear winners of optimization - the time gets spread out pretty quickly.  But it wouldn't hurt to have another set of eyes look at this.

<img width="1176" alt="screenshot_475" src="https://user-images.githubusercontent.com/3046552/89957608-155b5900-dbf5-11ea-84f1-b88bdc94d500.png">

Back up at the fork... 27.6% is spent in `stepSimulation.execute.(anonymous).emit`.  Drilling down, I see significant time spent in TinyEmitter.removeListener (14.3%) and phetioEngine.getAllRelatedTypes (8.6%).  See screenshot below.  I'd like to take a look at this with the iO team.

<img width="1174" alt="screenshot_474" src="https://user-images.githubusercontent.com/3046552/89957422-98c87a80-dbf4-11ea-84e7-3143fa4f4faf.png">

**Comment by pixelzoom** [comment]
Notes from Slack discussion with @samreid about the `stepSimulation.execute.(anonymous).emit` branch above:

* `TinyEmitter.removeListener` is currently O(N).  If it's converted to Set, it will be O(1) and will get rid of its 14.3%.

* `phetioEngine.getAllRelatedTypes` could definitely use some optimization. It's using `for in` which may not be the most efficient iterator.  It uses concat/spread operator and map, including a spread inside the `for in` loop.  And it calls itself recursively.

**Comment by pixelzoom** [comment]
When I drill down into `stepSimulation.execute.(anonymous).step`, I see lots of occurrences of `initializePhetioObject`, each one being only 1% or 2%.  But looking at Bottom-Up view, they add up to a whopping 16.4% (see screenshot below). Perhaps there's something that can be optimized there. 

<img width="647" alt="screenshot_477" src="https://user-images.githubusercontent.com/3046552/89959222-1abaa280-dbf9-11ea-9c8a-5aa38fa66737.png">

Looking more closely at Bottom-Up view...  I also see the following that might be investigated:

7% `phetioEngine.phetioObjectAdded`
3.5% `phetioEngine.ancestorMatches`

**Comment by pixelzoom** [comment]
`PhetioObject.initializePhetioObject` contains 14 calls to `validate`, which are no-ops when `assert` is disabled.  Why are these not short-circuited, so that we aren't calling a no-op function 14x for every PhetioObject?  If I comment out the 14 calls to `validate`, `initializePhetioObject` drops from 16.4% to 5.5% in Bottom-Up view:

<img width="851" alt="screenshot_478" src="https://user-images.githubusercontent.com/3046552/89959889-a84ac200-dbfa-11ea-9f03-9bcfc8459c20.png">

**Comment by pixelzoom** [comment]
I'm going to label this issue as "on hold" until the bottlenecks identified above are investigated by the iO team.

**Comment by samreid** [comment]
The above commit takes the self time of `ancestorMatches` from 5.3% down to 2.5%.

**Comment by samreid** [comment]
Changing `getAllRelatedTypes` to use `Array.prototype.push.apply( relatedTypes, typeIO.parameterTypes )` instead of `relatedTypes.push( ...typeIO.parameterTypes )` only shaves down its self time from 2.8% to 2.6%

**Comment by samreid** [comment]
After the commit in the issue and the commit in https://github.com/phetsims/tandem/issues/200, TinyEmitter.removeListener self-time is taking 20%.  That seems like the next part to work on.

**Comment by samreid** [comment]
https://github.com/phetsims/tandem/issues/200 has been addressed and the time for `TinyEmitter.removeListener` is close to 0%.  @pixelzoom can you please see if this is fast enough or if we should look for the next big thing?

**Comment by pixelzoom** [comment]
@samreid reported performance tests for this issue in https://github.com/phetsims/natural-selection/issues/140#issuecomment-676604571:
> I ran this on my iPad7 with a built version, running standalone in PhET-iO brand, and ?showTimes, and it reported:
> 
> time to mate: 348ms
time to Start Over: 184ms
> 
> On a Lenovo N23 Yoga Chromebook
> 
> time to mate: 1227ms
time to Start over: 738ms

@zepumph reported performance tests for this issue in https://github.com/phetsims/natural-selection/issues/140#issuecomment-676604571:
> On a build phet-io standalone off of master on an iPad7:
> 
> to go to the 6th gen: "time to mate = 547 ms"
startover: "time to Start Over = 210 ms".
> 
> second time (no refresh):
383ms
140ms
> 
> third time (no refresh):
296ms
137ms

**Comment by pixelzoom** [comment]
Following the testing steps in https://github.com/phetsims/natural-selection/issues/60#issuecomment-654325501...

For `brand=phet`, https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.12/phet/natural-selection_en_phet.html?showTimes&secondsPerGeneration=1:

| test device | OS | browser | times (ms) | average (ms) |
|:---|:---|:---| ---:|---:|
PixelZoom - iPad6 | iOS 13.6.1 | Safari 13 | 245, 159, 169 | 191 |

For `brand=phet-io`, https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.12/phet-io/natural-selection_all_phet-io.html?postMessageOnError&phetioStandalone&showTimes&secondsPerGeneration=1:

| test device | OS | browser | times (ms) | average (ms) |
|:---|:---|:---| ---:|---:|
PixelZoom - iPad6 | iOS 13.6.1 | Safari 13 | 607, 631, 389 | 542 |

**Comment by pixelzoom** [comment]
So based on the tests from @samreid @zepumph and me above, it looks like mating performance is acceptable on these devices, with the possible exception of @samreid's Lenovo N23 Yoga Chromebook (1227 ms).

@amanda-phet @ariel-phet while we established that < 1 second was acceptable performance for the "Start Over" button in #140, we have never established what is "acceptable" for the pause that occurs for mating.  What is the largest pause that you're willing to live with in a production version? 

Next step is to test on Lovelace Chromebook and any other test devices known to be "performance challenged".  @ariel-phet what is the availability of test resources for this task?

**Comment by ariel-phet** [comment]
@phet-steele please perform this test as well

@pixelzoom I think we can live with a fairly long pause for mating. Certainly what I am seeing on my own device is not even noticeable. If an old chromebook has a pause of 1 second or so between generations, I don't think it really does any pedagogical damage.

**Comment by phet-steele** [comment]
Please see https://github.com/phetsims/natural-selection/issues/140#issuecomment-676747493 for results of this test, but also I will just copy the results here to save everyone from clicking a link:

Following the testing steps in https://github.com/phetsims/natural-selection/issues/140#issuecomment-665288406 with 1.1.0-dev.12 ...

For `brand=phet`, https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.12/phet/natural-selection_en_phet.html?showTimes&secondsPerGeneration=1

| test device | OS | browser | times (ms) | average (ms) |
|:---|:---|:---| ---:|---:|
Feynman - Chromebook | 13099.62.0 | 84.0.4147.83 | Mate times : 730, 416, 400, (refreshed sim here), 830, 809, 492 | 613 |
| | | | Start Over times : 260, 261, 189, (refreshed sim here), 283, 247, 220 | 243 |

For `brand=phet-io`, https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.12/phet-io/natural-selection_all_phet-io.html?postMessageOnError&phetioStandalone&showTimes&secondsPerGeneration=1

| test device | OS | browser | times (ms) | average (ms) |
|:---|:---|:---| ---:|---:|
Feynman - Chromebook | 13099.62.0 | 84.0.4147.83 | Mate times : 1188, 938, 909, (refreshed sim here), 1206, 851, 765 | 976 |
| | | | Start Over times : 612, 405, 375, (refreshed sim here), 599, 425, 363 | 463 |

**Comment by pixelzoom** [comment]
In the above commit, I made the "time to mate" display a permanent feature of the sim, enabled via `?showTimes`. I had been planning to remove it, but it seems like this could be useful in the future.

**Comment by amanda-phet** [comment]
The absolute max time for mating should be capped at 3.3 seconds, right? That is the time that wolves come out, and we would need time to complete mating before that happens. To be safe, I think by 1.67 seconds the mating needs to be complete, and ideally I think 1s should be our target (for a worst-case scenario). 

Can someone test Lovelace, or is Feynman "performance challenged" enough to sign off on this?

**Comment by pixelzoom** [comment]
@amanda-phet I don't follow on how the cap is relevant to when the wolves come out.  When the sim pauses to mate, time stops, it doesn't continue.  So what's relevant is how long  a pause we are willing to tolerate.

**Comment by amanda-phet** [comment]
>@amanda-phet I don't follow on how the cap is relevant to when the wolves come out. When the sim pauses to mate, time stops, it doesn't continue. So what's relevant is how long a pause we are willing to tolerate.

Thanks for clarifying. I wondered if that was the case, but couldn't really test it since the performance has improved so much since I last saw the lag. I just thought I remembered the pause happening, and then seeing a "jump" ahead in the clock to catch up. So I wouldn't want that jump ahead to be too far. It's a bit fuzzy since I haven't seen the lag in a while. 

So, in the case when the sim pauses for let's say 800ms, does the clock stay at 12:00 and then continue moving ahead once the pause is over? Or will we see a similar "jump" ahead  to something like 1:00 once the pause ends?

**Comment by pixelzoom** [comment]
Ah, you know... @amanda-phet might be right about this.  If the sim pauses, we could get a really big `dt` from the browser that would cause the sim to jump ahead.  All environmental factors will still get applied correctly, but you might not see the wolves roaming around for their full length of time.  

Perhaps we should limit dt so that it's never greater than `secondsPerGeneration/N` where we decide what N should be.  So for example for `secondsPerGeneration=10`, if N=1, then we'll never advance the clock by more than 1 second in any timestep.

**Comment by pixelzoom** [comment]
> Perhaps we should limit dt ....

Never mind - we're already doing this, with N = 10.   See `GenerationClock.constrainDt`.

**Comment by amanda-phet** [comment]
> > Perhaps we should limit dt ....
>
> Never mind - we're already doing this, with N = 10. See GenerationClock.constrainDt.


OK, so I think my concern is not really a concern. 

How long of a pause are we willing to tolerate? How about 1s? Why not be consistent with the reset-all button.

**Comment by pixelzoom** [comment]
In the case of the "Start Over" button, the user is directly interacting with the thing that is causing the delay.  So laggy responsiveness is less tolerable.  For mating, the user may actually be passively watching (or interacting with something else), so we can get away with a longer pause.   1s may be a little more difficult to hit for this feature, but we seem to be pretty close on @phet-steele's Chromebook.  I guess we'll wait and see how it does on some of PhET's more challenged devices.

**Comment by samreid** [comment]
In https://github.com/phetsims/phet-io/issues/1695 we added caching to `phetioEngine.getAllRelatedTypes` which speeds up the generation 6 start time by about 5.5%.

**Comment by pixelzoom** [comment]
Excellent. I'll take my iPad6 for another performance test sometime early next week.

**Comment by pixelzoom** [comment]
For testing the 5.5% improvement mentioned by @samreid in https://github.com/phetsims/natural-selection/issues/60#issuecomment-677984085...

Following the testing steps in https://github.com/phetsims/natural-selection/issues/60#issuecomment-654325501 with 1.1.0-dev.12 ...

For `brand=phet`, https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.14/phet/natural-selection_en_phet.html?showTimes&secondsPerGeneration=1

| test device | OS | browser | times (ms) | average (ms) | previous (ms) |
|:---|:---|:---| ---:|---:|---:|
PixelZoom - iPad6 | iOS 13.6.1 | Safari 13 |  102, 95, 91 | 96 | 191 |
PixelZoom - 2019 MacBook Pro | macOS 10.15.6 | Chrome 84.0.4147.135 |  72, 62, 60 | 65  | 139 |

For `brand=phet-io`, https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.14/phet-io/natural-selection_all_phet-io.html?postMessageOnError&phetioStandalone&showTimes&secondsPerGeneration=1

| test device | OS | browser | times (ms) | average (ms) | previous (ms) |
|:---|:---|:---| ---:|---:|---:|
PixelZoom - iPad6 | iOS 13.6.1 | Safari 13 | 191, 125, 128 |  148 | 542 |
PixelZoom - 2019 MacBook Pro | macOS 10.15.6 | Chrome 84.0.4147.135 |  115, 109, 92 | 105 | 202 |

This is way more than a 5.5% improvement.  The PhET-iO version is running 3.5x faster on iPad, and 2x on MacBook Pro.  @samreid were there other improvements made that I may have picked up?  My most recent pull-all was ~10am this morning.

**Comment by pixelzoom** [comment]
Oh rats (bunnies?) I was testing the wrong thing here.  I was testing performance of "Start Over", and this issue is about mating.  Starting over...

**Comment by pixelzoom** [comment]
Over in https://github.com/phetsims/natural-selection/issues/140#issuecomment-679396378, @samreid said:
> I think `getAllRelatedTypes` is called during object creation, so more likely related to #60 which is about bunny creation.

So a 5.5% performance increase is indeed expected for this issue.  I don't see that increase, and in fact see a huge _decrease_ in performance (2x slower) on iPad6.   Thoughts?

**Comment by samreid** [comment]
Some notes from recent runs:

Sam iPad 6
OS 13.3
Time to mate: 396ms, 391ms, 392ms, 372ms
Start over: 175ms, 141ms
Time to mate: 332ms, 331ms

Chris 13.6.1
590, 757, 757, 452, 348

**Comment by pixelzoom** [comment]
Testing 1.1.0-dev.16 brand=phet-io again on my iPad6, ~15 runs were all in the 450-600 ms range.  I'm not sure what was going on with the iPad6 results in https://github.com/phetsims/natural-selection/issues/60#issuecomment-679326446,  but I guess I won't worry about it.

**Comment by pixelzoom** [comment]
On a Lenovo 300e Chromebook, @samreid reports for 1.1.0-dev.16 brand=phet-io:

time to mate = 1131
time to Start Over = 621 ms

That seems reasonable to me.

**Comment by pixelzoom** [comment]
@samreid says there are no additional PhET-iO performance improvements planned.  

Performance now seems acceptable to me, including on lower end devices like Chromebooks.  So I think we're ready to move to the performance "sign-off" phase in #191.

**Comment by KatieWoe** [comment]
@ariel-phet asked for the https://github.com/phetsims/natural-selection/issues/60#issuecomment-679326446 tests to be done on Lovelace since it is a slower system. Results:

| test device | OS | test | times (ms) | average (ms) |
|:---|:---|:---| ---:|---:|
Lovelace | ChromeOS 85.0.4183.59 | phet brand time to mate | 1073, 656, 545 | 758 |
Lovelace | ChromeOS 85.0.4183.59 | phet brand time to start over | 427, 407, 327 | 387 |
Lovelace | ChromeOS 85.0.4183.59 | phet io time to mate | 1541, 1250, 1092 | 1294 |
Lovelace | ChromeOS 85.0.4183.59 | phet io time to start over | 860, 710, 859 | 810 |

Note that the bug where clicking "Start Over" twice skips the button to add a friend. This seems like it would be relatively easy to do on even a faster moving device.

**Comment by pixelzoom** [comment]
@KatieWoe said:
> Note that the bug where clicking "Start Over" twice skips the button to add a friend. This seems like it would be relatively easy to do on even a faster moving device.

I brought this usability issue up in https://github.com/phetsims/natural-selection/issues/166.  @amanda-phet didn't see evidence of this in interviews, and she decided it's not a problem.  I'm going to re-open that issue, add @KatieWoe's comment there, and let @amanda-phet revisit.

**Comment by amanda-phet** [comment]
@pixelzoom said
>Performance for Lovelace looks reasonable to me. "Start Over" is < 1 second, and time to mate is not huge. 

It looks like time to mate with PhET-iO is kind of huge (up to 1541 ms), but that seems like the worst case scenario. I am totally fine with the performance in all other areas, and since I can't really experience the 1.5s pause I'll just have to defer to you all if that seems acceptable. 

I'll respond to #166 directly. Closing this issue.

**Comment by pixelzoom** [comment]
I've have a couple new ideas about this issue that I'd like to bounce off of @amanda-phet.  Reopening and assigning to me so that I remember.

**Comment by pixelzoom** [comment]
@amanda-phet and I discussed a couple of other idea that I had:

(1) Create bunnies in small batches, say 50 at a time. This would still cause the clock to pause at 12:00 (and maybe even longer), but you'd see something happening during that pause.  The downside is that this isn't a good match for the "single data point" that mating produces.

(2) Pool (reuse) bunny SpriteInstances. Unlikely to have a big performance impact, likely to make the code more complicated (and it's already complicated). 

So we decided to leave this where it is, and live with the brief pause at 12:00.

**Comment by pixelzoom** [comment]
A few more notes about:
> (2) Pool (reuse) bunny SpriteInstances. ....

I investigate this approach briefly with @jonathanolson, and it was pretty easy to get pooling working.  But pooling turned out to be a wash (didn't change the performance) and was more complicated, so I opted to keep things as is.

But this exercise did result in improvements.  @jonathanolson identified some inefficiencies (use of `merge`, object allocations,...) in my SpriteInstances class hierarchy.  I improved those things in the above commit, and I'm seeing ~35% performance improvement for "time to mate" on my MacBook Pro (unbuilt brand=phet version in Chrome).  Time to mate for the 6th generation dropped from 130ms to 85ms.
